File: Copy-List-With-Random-Pointer.cpp
// Time Complexity : O(3N)

// Space Complexity : O(1) - since the answer is stored as a LL

// Did this code successfully run on Leetcode : Yes

// Appoarch: 3 main steps:
// 1. Create a new node next to the original node
// 2. Create random connections between them
// 3. Connect next pointers by disconnecting new ones. 

// 138. Copy List with Random Pointer

#include <bits/stdc++.h>
using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};

class Solution {
public:
    // Create new nodes in between
    void insertCopyInBetween(Node* head){
        Node* temp = head;
        while(temp != NULL){
            Node* nextLink = temp->next;
            Node* copyNode = new Node(temp->val);
            copyNode->next = nextLink;
            temp->next = copyNode;
            temp = nextLink;
        }
    }
    // create random connections
    void connectRandomPointers(Node* head){
        Node* temp = head;
        while(temp != NULL){
            Node* copyNode = temp->next;
            if(temp->random){
                copyNode->random = temp->random->next;
            }
            else{
                copyNode->random = NULL;
            }
            temp = temp->next->next;
        }
    }
    // disconnect new connections to get back old LL
    Node* getDeepCopyList(Node* head){
        Node* temp = head;
        Node* dNode = new Node(-1);
        Node* res = dNode;
        while(temp != NULL){
            // creating new list
            res->next = temp->next;
            res = res->next;
            // disconnecting new link to get back original LL
            temp->next = temp->next->next;
            temp = temp->next;
        }
        return dNode->next;
    }

    Node* copyRandomList(Node* head) {
        insertCopyInBetween(head);
        connectRandomPointers(head);
        return getDeepCopyList(head);
    }
};

File: CopyListWithRandomPointer.cs
using System;
using System.Collections.Generic;


namespace Algorithms
{
    public class Node
    {
        public int val;
        public Node next;
        public Node random;

        public Node(int _val)
        {
            val = _val;
            next = null;
            random = null;
        }
    }
    /// Time Complexity : O(n)
    // Space Complexity :O(n)
    // Did this code successfully run on Leetcode :Yes
    // Any problem you faced while coding this :  Referred Online
    public class CopyListWithRandomPointer
    {
        Dictionary<Node, Node> map;
        public Node CopyRandomList(Node head)
        {
            if (head == null) return head;
            map = new Dictionary<Node, Node>();
            Node deepHead = clone(head);
            Node currCopy = deepHead;
            Node curr = head;
            while (curr != null)
            {
                currCopy.next = clone(curr.next);
                currCopy.random = clone(curr.random);
                curr = curr.next;
                currCopy = currCopy.next;
            }
            return deepHead;
        }

        private Node clone(Node node)
        {
            if (node == null) return null;
            if (map.ContainsKey(node))
            {
                return map.GetValueOrDefault(node);
            }
            Node newNode = new Node(node.val);
            map.Add(node, newNode);
            return newNode;
        }


    }
}
